modules/web-console/frontend/app/components/page-queries/components/queries-notebook/controller.ts

```diff
@@ -1233,46 +1264,109 @@ export class NotebookCtrl {
             _rebuildColumns(paragraph);
         };
 
-        const _showLoading = (paragraph, enable) => {
-            if (paragraph.qryType === 'scan')
-                paragraph.scanningInProgress = enable;
-
-            paragraph.loading = enable;
-        };
-
-        const _fetchQueryResult = (paragraph, clearChart, res, qryArg) => {
-            if (!_.isNil(res.rows)) {
-                _processQueryResult(paragraph, clearChart, res);
-                _tryStartRefresh(paragraph);
-
-                $scope.$applyAsync();
+        /**
+         * Execute query and get first result page.
+         *
+         * @param qryType Query type. 'query' or `scan`.
+         * @param qryArg Argument with query properties.
+         * @param {(res) => any} onQueryStarted Action to execute when query ID is received.
+         * @return {Observable<VisorQueryResult>} Observable with first query result page.
+         */
+        const _executeQuery0 = (qryType, qryArg, onQueryStarted: (res) => any = () => {}) => {
+            return from(qryType === 'scan' ? agentMgr.queryScan(qryArg) : agentMgr.querySql(qryArg)).pipe(
+                tap((res) => {
+                    onQueryStarted(res);
+                    $scope.$applyAsync();
+                }),
+                expand((res) => {
```

Comment:
I think that an expand, filter  is not really needed and can be replaced with something like this: 
```
exhaustMap((res) => {
   if (!_.isNil(res.rows))
      return from(res);

   return merge(fetchFirstPageTask, pingQueryTask);
}) 
```

@Klaster1 what do you think about this?

Reply:
Yeah, I'm not sure recursion is even require here. The `exhaust` block basically polls first page results and pings to keep the polling alive. If I understand the query API correctly, `res.queryId` - the only value received from previous recursion call - does not change ever.

Reply:
Recursion is removed by changing of exexpand to exhaustMap.